### PR TITLE
refactor validate block leader so we can not special case rule errors

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* feature/pos-fast-hotstuff-consensus
  * **PR #742** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/742" target="_blank">Graphite</a> 
    * **PR #783** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/783" target="_blank">Graphite</a>  👈
      * **PR #784** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/784" target="_blank">Graphite</a> 
        * **PR #776** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/776" target="_blank">Graphite</a> 
          * **PR #777** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/777" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/783?utm_source=stack-comment).

File: lib/pos_blockchain.go [hunk 1]
```diff
@@ -413,33 +410,35 @@ func (bc *Blockchain) validateBlockView(desoBlock *MsgDeSoBlock) error {
 }
 
 // validateBlockLeader validates that the proposer is the expected proposer for the
-// block height + view number pair.
-func (bc *Blockchain) validateBlockLeader(desoBlock *MsgDeSoBlock) error {
-	utxoView, err := NewUtxoView(bc.db, bc.params, bc.postgres, bc.snapshot)
+// block height + view number pair. It returns a bool indicating whether or not
+// we confirmed that the leader is invalid. If we receive an error, we are unsure
+// if the leader is invalid or not, so we return false.
+func (bc *Blockchain) validateBlockLeader(block *MsgDeSoBlock) (_isValidateFailed bool, _err error) {
```
Comment (tholonious) [inline lib/pos_blockchain.go:416]: nit: use `isXYZ` or `hasXYZ` naming for these types of functions. For consistency with the function name, it'll be good to make the boolean return value match the true/false meaning of the function
```suggestion
func (bc *Blockchain) hasValidBlockProposerPoS(block *MsgDeSoBlock) (hasValidBlockProposer bool, _err error) {
```

File: lib/pos_blockchain.go [hunk 2]
```diff
@@ -197,6 +190,14 @@ func (bc *Blockchain) validateAndIndexBlockPoS(block *MsgDeSoBlock) (*BlockNode,
 		return bc.storeValidateFailedBlockWithWrappedError(block, err)
 	}
 
+	isInvalidLeader, err := bc.validateBlockLeader(block)
```
Comment (tholonious) [inline lib/pos_blockchain.go:193]: nit
```suggestion
	isBlockProposerValid, err := bc.hasValidBlockProposerPoS(block)
```

File: lib/pos_blockchain.go [hunk 3]
```diff
@@ -197,6 +190,14 @@ func (bc *Blockchain) validateAndIndexBlockPoS(block *MsgDeSoBlock) (*BlockNode,
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem validating block leader")
+	}
+	if isInvalidLeader {
```
Comment (tholonious) [inline lib/pos_blockchain.go:197]: nit
```suggestion
	if ! isBlockProposerValid {
```

**Review by tholonious — CHANGES_REQUESTED** [review]

**Review by tholonious — APPROVED** [review]

**Pull request merged**